### DISCUSSION TITLE
how use Sample client GFX_DVC?

### DISCUSSION THREAD

**Discussion by kimcoldwater** [FreeRDP/FreeRDP#8345] [Q&A]
Hello, I am creating a project using a sample client.

When using the command line option `/gfx` on the sample client, `invalid->width = 1` , `invalid->height = 1` are continuously output from `end_paint`.

When `gfx_dvc` is not used, data normally comes into `end_paint`

There is `gdi_graphics_pipeline_init` specified in the channel list, is there anything I need to implement separately?

I give /gdi:sw and /bpp:32 options.

Is there anything I missed?

**Comment by akallabeth** [comment]
well, a bit more information would help, can´t reproduce the errors you mention and the sample client normally connects (`master` and `stable-2.0` branch)

**Reply by kimcoldwater** [reply]
ok.

I added the `-DWITH_DEBUG_RDPGFX=ON` option.

And it's my build option.

```
BUILD_TESTING=OFF BUILTIN_CHANNELS=ON HAVE_AIO_H=1 HAVE_EXECINFO_H=1 HAVE_FCNTL_H=1 HAVE_GETLOGIN_R=1 HAVE_GETPWUID_R=1 HAVE_INTTYPES_H=1 HAVE_MATH_C99_LONG_DOUBLE=1 HAVE_POLL_H=1 HAVE_PTHREAD_MUTEX_TIMEDLOCK=ON HAVE_PTHREAD_MUTEX_TIMEDLOCK_LIB=1 HAVE_PTHREAD_MUTEX_TIMEDLOCK_SYMBOL= HAVE_SYSLOG_H=1 HAVE_SYS_EVENTFD_H=1 HAVE_SYS_FILIO_H= HAVE_SYS_MODEM_H= HAVE_SYS_SELECT_H=1 HAVE_SYS_SOCKIO_H= 
HAVE_SYS_STRTIO_H= HAVE_SYS_TIMERFD_H=1 HAVE_TM_GMTOFF=1 HAVE_UNISTD_H=1 WITH_ALSA=ON WITH_CAIRO=OFF WITH_CCACHE=ON WITH_CHANNELS=ON WITH_CLANG_FORMAT=ON WITH_CLIENT=ON WITH_CLIENT_AVAILABLE=1 WITH_CLIENT_CHANNELS=ON WITH_CLIENT_CHANNELS_AVAILABLE=1 WITH_CLIENT_COMMON=ON WITH_CLIENT_INTERFACE=OFF WITH_CUPS=OFF WITH_DEBUG_ALL=OFF WITH_DEBUG_CAPABILITIES=OFF WITH_DEBUG_CERTIFICATE=OFF WITH_DEBUG_CHANNELS=OFF WITH_DEBUG_CLIPRDR=OFF WITH_DEBUG_DVC=OFF WITH_DEBUG_KBD=OFF WITH_DEBUG_LICENSE=OFF WITH_DEBUG_MUTEX=OFF WITH_DEBUG_NEGO=OFF 
WITH_DEBUG_NLA=OFF WITH_DEBUG_NTLM=OFF WITH_DEBUG_RAIL=OFF WITH_DEBUG_RDP=OFF WITH_DEBUG_RDPDR=OFF WITH_DEBUG_RDPEI=OFF WITH_DEBUG_RDPGFX=ON WITH_DEBUG_REDIR=OFF WITH_DEBUG_RFX=OFF WITH_DEBUG_RINGBUFFER=OFF WITH_DEBUG_SCARD=OFF WITH_DEBUG_SND=OFF WITH_DEBUG_SVC=OFF WITH_DEBUG_SYMBOLS=OFF WITH_DEBUG_THREADS=OFF WITH_DEBUG_TIMEZONE=OFF WITH_DEBUG_TRANSPORT=OFF WITH_DEBUG_TSG=OFF WITH_DEBUG_TSMF=OFF WITH_DEBUG_TSMF=OFF WITH_DEBUG_TSMF_AVAILABLE=0 WITH_DEBUG_WND=OFF WITH_DEBUG_X11=OFF WITH_DEBUG_X11_CLIPRDR=OFF WITH_DEBUG_X11_LOCAL_MOVESIZE=OFF 
WITH_DEBUG_XV=OFF WITH_DSP_EXPERIMENTAL=OFF WITH_EVENTFD_READ_WRITE=1 WITH_FAAC=OFF WITH_FAAD2=OFF WITH_FFMPEG=OFF WITH_GFX_H264=OFF WITH_GPROF=OFF WITH_GSM=OFF WITH_GSSAPI=OFF WITH_ICU=OFF WITH_IPP=OFF WITH_JPEG=OFF WITH_LAME=OFF WITH_LIBRARY_VERSIONING=ON WITH_LIBSYSTEMD=OFF WITH_MACAUDIO=OFF WITH_MACAUDIO=OFF 
WITH_MACAUDIO_AVAILABLE=0 WITH_MANPAGES=ON WITH_MBEDTLS=OFF WITH_OPENCL=OFF WITH_OPENH264=OFF WITH_OPENSLES=OFF WITH_OPENSSL=ON WITH_OSS=ON WITH_PCSC=OFF WITH_PROFILER=OFF WITH_PULSE=OFF WITH_SAMPLE=ON WITH_SANITIZE_ADDRESS=OFF WITH_SANITIZE_ADDRESS_AVAILABLE=1 WITH_SANITIZE_MEMORY=OFF WITH_SANITIZE_MEMORY_AVAILABLE=1 WITH_SANITIZE_THREAD=OFF WITH_SANITIZE_THREAD_AVAILABLE=1 WITH_SERVER=OFF WITH_SERVER_INTERFACE=ON WITH_SMARTCARD_INSPECT=OFF WITH_SOXR=OFF WITH_SSE2=ON WITH_SWSCALE=OFF WITH_THIRD_PARTY=OFF WITH_VALGRIND_MEMCHECK=OFF WITH_VALGRIND_MEMCHECK_AVAILABLE=1
 WITH_VERBOSE_WINPR_ASSERT=ON WITH_WAYLAND=OFF WITH_WINPR_TOOLS=ON WITH_X11=OFF WITH_XEXT=OFF WITH_ZLIB=ON
Build type:          Release
CFLAGS:               -fPIC -Wall -Wno-unused-result -Wno-unused-but-set-variable -Wno-deprecated-declarations -fvisibility=hidden -Wimplicit-function-declaration -Wredundant-decls -fno-omit-frame-pointer -DWINPR_DLL
Compiler:            GNU, 7.3.1
Target architecture: x64
```


Below is the debug log that came on when the time changed in the lower right corner of the window screen.


```
[16:07:00:917] [14798:14865] [DEBUG][com.freerdp.channels.drdynvc.client] - order_recv: Cmd=0x3, Sp=1 cbChId=0
[16:07:00:917] [14798:14865] [DEBUG][com.freerdp.channels.rdpgfx.client] - cmdId: RDPGFX_CMDID_STARTFRAME (0x000B) flags: 0x0000 pduLength: 16
[16:07:00:917] [14798:14865] [DEBUG][com.freerdp.channels.rdpgfx.client] - RecvStartFramePdu: frameId: 51 timestamp: 0x0040B53B
[16:07:00:917] [14798:14865] [DEBUG][com.freerdp.channels.rdpgfx.client] - cmdId: RDPGFX_CMDID_WIRETOSURFACE_1 (0x0001) flags: 0x0000 pduLength: 255
[16:07:00:917] [14798:14865] [DEBUG][com.freerdp.channels.rdpgfx.client] - RecvWireToSurface1Pdu: surfaceId: 1 codecId: RDPGFX_CODECID_CLEARCODEC (0x0008) pixelFormat: 0x20 destRect: left: 933 top: 734 right: 940 bottom: 743 bitmapDataLength: 230
[16:07:00:917] [14798:14865] [DEBUG][com.freerdp.channels.rdpgfx.client] - cmdId: RDPGFX_CMDID_ENDFRAME (0x000C) flags: 0x0000 pduLength: 12
[16:07:00:917] [14798:14865] [DEBUG][com.freerdp.channels.rdpgfx.client] - RecvEndFramePdu: frameId: 51
[16:07:00:917] [14798:14865] [DEBUG][com.freerdp.channels.rdpgfx.client] - SendFrameAcknowledgePdu: 51
[16:07:00:917] [14798:14865] [DEBUG][com.freerdp.channels.drdynvc.client] - write_data: ChannelId=7 size=20
```

I planted Log in end_paint, but it was not called.

The gfx works normally, but do I have to do something different?

**Reply by akallabeth** [reply]
which version are you testing with? on `master` the sample includes `FreeRDP_DeactivateClientDecoding` which does exactly what the name suggests (e.g. disable graphics decoding)

**Reply by kimcoldwater** [reply]
i use 2.8.0 release . version

**Reply by akallabeth** [reply]
* you get a (`gdi_EndFrame`) `EndFrame` callback in `RDPGFX_CMDID_ENDFRAME` (check `rdpgfx_recv_end_frame_pdu`)
* That calls `UpdateSurfaces` (defaults to `gdi_UpdateSurfaces`)
* that calls `EndPaint`

**Reply by akallabeth** [reply]
so, unless you have  `surface->outputMapped = FALSE` set or  `gdi->suppressOutput` set you get the callback?

**Reply by kimcoldwater** [reply]
Thanks to your kind advice, I easily found and understood the key part.

Then, where should I put the setting value?
`postconnect`? `begin_paint`?
 Can you put the setting value in the client source without touching the core?

**Reply by akallabeth** [reply]
@kimcoldwater what do you mean? if you don´t set anything it will work.

**Reply by kimcoldwater** [reply]
I didn't touch any gfx related settings in the sample client configuration.

Of course, the two options you mentioned above.

I just added `/gfx` to the command line option.

If you insert `/gfx` by itself, `end_paint` is not called, but if you remove it, it is called properly.

I don't touch anything related to graphics, I just use data after connecting

**Reply by akallabeth** [reply]
@kimcoldwater not what I can see here.

added
```
static uint64_t xxx = 0;
    char filename[8192]={0};
    snprintf(filename, sizeof(filename), "/tmp/image-%08" PRIx64 ".bmp", xxx++);
    winpr_bitmap_write(filename, gdi->primary_buffer, gdi->width, gdi->height, GetBitsPerPixel( gdi->dstFormat));
    freerdp_input_send_keyboard_event_ex(context->input, TRUE, RDP_SCANCODE_LWIN);
    freerdp_input_send_keyboard_event_ex(context->input, FALSE, RDP_SCANCODE_LWIN);
```

in `tf_end_paint` and it will always update until I end the client connection. (note, the bitmap is not really usable/distorted as the image stride does not match width exactly)

**Reply by kimcoldwater** [reply]
All right.
I'll check the rest while testing it.
Thank you very much for your kind reply.

**Reply by kimcoldwater** [reply]
`if (!freerdp_settings_set_bool(context->settings, FreeRDP_DeactivateClientDecoding, FALSE))`

The `/gfx` option worked fine when the item was disabled.

```
	/* With this setting we disable all graphics processing in the library.
	 *
	 * This allows low resource (client) protocol parsing.
```
With these comments, what does it mean to disable all graphics processing?

I'm implementing a headless client.

Which would help improve performance, setting this to TRUE or using /gfx? High resolution of 1920x1080 or higher is also available.

**Reply by akallabeth** [reply]
just what the comment tells, it will not decode any graphics client side (e.g. you will not see any valid data in `EndPaint`)
